flink-tests/src/test/java/org/apache/flink/test/checkpointing/SnapshotFileMergingCompatibilityITCase.java

```diff
@@ -0,0 +1,126 @@
+            RestoreMode restoreMode, boolean fileMergingAcrossBoundary, @TempDir Path checkpointDir)
+            throws Exception {
+        testSwitchingFileMerging(
+                checkpointDir, true, false, restoreMode, fileMergingAcrossBoundary);
+    }
+
+    private void testSwitchingFileMerging(
+            Path checkpointDir,
+            boolean firstFileMergingSwitch,
+            boolean secondFileMergingSwitch,
+            RestoreMode restoreMode,
+            boolean fileMergingAcrossBoundary)
+            throws Exception {
+        final Configuration config = new Configuration();
+        config.set(CheckpointingOptions.CHECKPOINTS_DIRECTORY, checkpointDir.toUri().toString());
+        config.set(CheckpointingOptions.INCREMENTAL_CHECKPOINTS, true);
+        config.set(CheckpointingOptions.FILE_MERGING_ACROSS_BOUNDARY, fileMergingAcrossBoundary);
+        config.set(CheckpointingOptions.FILE_MERGING_ENABLED, firstFileMergingSwitch);
+        MiniClusterWithClientResource firstCluster =
+                new MiniClusterWithClientResource(
+                        new MiniClusterResourceConfiguration.Builder()
+                                .setConfiguration(config)
+                                .setNumberTaskManagers(2)
+                                .setNumberSlotsPerTaskManager(2)
+                                .build());
+        EmbeddedRocksDBStateBackend stateBackend1 = new EmbeddedRocksDBStateBackend();
+        stateBackend1.configure(config, Thread.currentThread().getContextClassLoader());
+        firstCluster.before();
+        String externalCheckpoint;
+        try {
+            externalCheckpoint =
+                    runJobAndGetExternalizedCheckpoint(
+                            stateBackend1, null, firstCluster, restoreMode, config);
+            assertThat(externalCheckpoint).isNotNull();
+        } finally {
+            firstCluster.after();
+        }
+
+        config.set(CheckpointingOptions.FILE_MERGING_ENABLED, secondFileMergingSwitch);
+        EmbeddedRocksDBStateBackend stateBackend2 = new EmbeddedRocksDBStateBackend();
+        stateBackend2.configure(config, Thread.currentThread().getContextClassLoader());
+        MiniClusterWithClientResource secondCluster =
+                new MiniClusterWithClientResource(
+                        new MiniClusterResourceConfiguration.Builder()
+                                .setConfiguration(config)
+                                .setNumberTaskManagers(2)
+                                .setNumberSlotsPerTaskManager(2)
+                                .build());
+        secondCluster.before();
+        try {
+            String secondCheckpoint =
+                    runJobAndGetExternalizedCheckpoint(
+                            stateBackend2, externalCheckpoint, secondCluster, restoreMode, config);
+            assertThat(secondCheckpoint).isNotNull();
+        } finally {
+            secondCluster.after();
+        }
```

Comment:
IIUC, should we do more check on the second checkpoint? For example, does a new job can restore from the second cp, indicating it is complete?

Reply:
Done.

Reply:
Seems you are taking another `thirdCheckpoint` without building another cluster. I mean, we should verify the availability of `secondCheckpoint`. So maybe build `thirdCluster` and restore from the `secondCheckpoint`? Also, it there any way to check the state handles to be `SegmentStreamStateHandle` or not?

Reply:
How about using `CheckpointMetadata metadata = TestUtils.loadCheckpointMetadata(firstCheckpoint);` and visit all the state handle within it?

And I found there is no managed keyed state in current tests, maybe you should fill some data in.

Reply:
> How about using CheckpointMetadata metadata = TestUtils.loadCheckpointMetadata(firstCheckpoint); and visit all the state handle within it?

Thanks for your suggestion! I have added the check logic in  SnapshotFileMergingCompatibilityITCase#verifyStateHandleType. PTAL again.

> And I found there is no managed keyed state in current tests, maybe you should fill some data in.

Actually, there are managed keyed states being generated in current tests. I confirmed this in debug mode and added the corresponding check in this UT case.